editor/src/clj/editor/editor_extensions/graph.clj

```diff
@@ -0,0 +1,154 @@
+;; https://www.defold.com/license
+;;
+;; Unless required by applicable law or agreed to in writing, software distributed
+;; under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
+;; specific language governing permissions and limitations under the License.
+
+(ns editor.editor-extensions.graph
+  "Code for interacting with graph from editor scripts, i.e. get/set values"
+  (:require [clojure.spec.alpha :as s]
+            [clojure.string :as string]
+            [dynamo.graph :as g]
+            [editor.code.util :as code.util]
+            [editor.defold-project :as project]
+            [editor.editor-extensions.validation :as validation]
+            [editor.outline :as outline]
+            [editor.properties :as properties]
+            [editor.resource :as resource]
+            [editor.types :as types])
+  (:import [clojure.lang MultiFn]
+           [org.luaj.vm2 LuaError]))
+
+(defn node-id-or-path->node-id
+  "Coerce a value that may be node id or proj-path to existing node id"
+  [node-id-or-path project evaluation-context]
+  (if (string? node-id-or-path)
+    (let [node-id (project/get-resource-node project node-id-or-path evaluation-context)]
+      (when (nil? node-id)
+        (throw (LuaError. (str node-id-or-path " not found"))))
+      node-id)
+    node-id-or-path))
+
+(defn node-id->type-keyword [node-id ec]
+  (g/node-type-kw (:basis ec) node-id))
+
+;; region get
+
+(defn- ensuring-converter [spec]
+  (fn [value _outline-property _project _evaluation-context]
+    (validation/ensure spec value)))
+
+(defn- resource-converter [node-id-or-path outline-property project evaluation-context]
+  (validation/ensure (s/or :nothing #{""} :resource ::validation/node-id) node-id-or-path)
+  (when-not (= node-id-or-path "")
+    (let [node-id (node-id-or-path->node-id node-id-or-path project evaluation-context)
+          resource (g/node-value node-id :resource evaluation-context)
+          ext (:ext (properties/property-edit-type outline-property))]
+      (when (seq ext)
+        (validation/ensure (set ext) (resource/type-ext resource)))
+      resource)))
+
+(def ^:private edit-type-id->value-converter
+  {g/Str {:to identity :from (ensuring-converter string?)}
+   g/Bool {:to identity :from (ensuring-converter boolean?)}
+   g/Num {:to identity :from (ensuring-converter number?)}
+   g/Int {:to identity :from (ensuring-converter int?)}
+   types/Vec2 {:to identity :from (ensuring-converter (s/tuple number? number?))}
+   types/Vec3 {:to identity :from (ensuring-converter (s/tuple number? number? number?))}
+   types/Vec4 {:to identity :from (ensuring-converter (s/tuple number? number? number? number?))}
+   'editor.resource.Resource {:to resource/proj-path :from resource-converter}})
+
+(defn- multi-responds? [^MultiFn multi & args]
+  (some? (.getMethod multi (apply (.-dispatchFn multi) args))))
```

Comment:
A candidate for the `fn` module, perhaps?

Reply:
Yes!